File: crates/fraiseql-core/src/validation/custom_scalar.rs
//! Custom scalar trait and validation for user-defined scalars.
//!
//! This module provides a trait-based system for defining custom GraphQL scalars
//! at runtime, allowing applications to implement their own validation logic.
//!
//! # Example
//!
//! ```
//! use fraiseql_core::validation::CustomScalar;
//! use fraiseql_core::error::{FraiseQLError, Result};
//! use serde_json::Value;
//!
//! #[derive(Debug)]
//! struct Email;
//!
//! impl CustomScalar for Email {
//!     fn name(&self) -> &str {
//!         "Email"
//!     }
//!
//!     fn serialize(&self, value: &Value) -> Result<Value> {
//!         Ok(value.clone())
//!     }
//!
//!     fn parse_value(&self, value: &Value) -> Result<Value> {
//!         let str_val = value.as_str()
//!             .ok_or_else(|| FraiseQLError::parse("expected string"))?;
//!
//!         if !str_val.contains('@') {
//!             return Err(FraiseQLError::validation(
//!                 format!("invalid email format: {}", str_val)
//!             ));
//!         }
//!
//!         Ok(Value::String(str_val.to_string()))
//!     }
//!
//!     fn parse_literal(&self, ast: &Value) -> Result<Value> {
//!         self.parse_value(ast)
//!     }
//! }
//!
//! let email = Email;
//! assert_eq!(email.name(), "Email");
//! ```

use std::fmt;

use serde_json::Value;

use crate::error::Result;

/// Trait for implementing custom GraphQL scalar types.
///
/// Implement this trait to create custom scalars with validation logic.
/// Each method represents a different validation context in GraphQL.
pub trait CustomScalar: Send + Sync + fmt::Debug {
    /// Returns the GraphQL scalar type name (e.g., "Email", "Phone").
    fn name(&self) -> &str;

    /// Serialize a database value to a GraphQL response value.
    ///
    /// This is called when returning values from resolvers.
    ///
    /// # Errors
    ///
    /// Returns `FraiseQLError::Validation` if the value cannot be serialized for this scalar type.
    fn serialize(&self, value: &Value) -> Result<Value>;

    /// Parse a variable value from a GraphQL operation.
    ///
    /// This is called when parsing variables in GraphQL operations.
    /// For example: `{ query: getUser($email: Email) }`
    ///
    /// # Errors
    ///
    /// Returns `FraiseQLError::Validation` if the value does not conform to this scalar's format.
    fn parse_value(&self, value: &Value) -> Result<Value>;

    /// Parse a literal value from a GraphQL query string.
    ///
    /// For example: `{ user(email: "test@example.com") }`
    /// The `ast` parameter is typically an object with a "value" key for simple types.
    ///
    /// # Errors
    ///
    /// Returns `FraiseQLError::Validation` if the literal does not conform to this scalar's format.
    fn parse_literal(&self, ast: &Value) -> Result<Value>;
}

/// Result of custom scalar validation.
pub type CustomScalarResult = Result<Value>;

File: sdks/official/fraiseql-go/fraiseql/custom_scalar.go
package fraiseql

import "fmt"

// CustomScalar defines the interface for custom GraphQL scalar types.
//
// Implement this interface to create custom scalars with validation logic.
// Each method represents a different validation context in GraphQL.
//
// Example:
//
//	type Email struct{}
//
//	func (e *Email) Name() string {
//	    return "Email"
//	}
//
//	func (e *Email) Serialize(value interface{}) (interface{}, error) {
//	    return value, nil
//	}
//
//	func (e *Email) ParseValue(value interface{}) (interface{}, error) {
//	    str := toString(value)
//	    if !strings.Contains(str, "@") {
//	        return nil, fmt.Errorf("invalid email format: %s", str)
//	    }
//	    return str, nil
//	}
//
//	func (e *Email) ParseLiteral(ast interface{}) (interface{}, error) {
//	    m, ok := ast.(map[string]interface{})
//	    if ok {
//	        if val, exists := m["value"]; exists {
//	            return e.ParseValue(val)
//	        }
//	    }
//	    return nil, fmt.Errorf("email literal must be string")
//	}
//
//	// Register the scalar
//	func init() {
//	    RegisterCustomScalar(&Email{})
//	}
type CustomScalar interface {
	// Name returns the GraphQL scalar type name (e.g., "Email", "Phone").
	Name() string

	// Serialize converts a database value to a GraphQL response value.
	// This is called when returning values from resolvers.
	Serialize(value interface{}) (interface{}, error)

	// ParseValue parses a variable value from GraphQL.
	// This is called when parsing variables in GraphQL operations.
	// For example: { query: getUser($email: Email) }
	ParseValue(value interface{}) (interface{}, error)

	// ParseLiteral parses a literal value from a GraphQL query string.
	// For example: { user(email: "test@example.com") }
	// The ast parameter is typically a map with a "value" key for simple types.
	ParseLiteral(ast interface{}) (interface{}, error)
}

// toString converts a value to string, handling common types.
func toString(value interface{}) string {
	switch v := value.(type) {
	case string:
		return v
	case []byte:
		return string(v)
	default:
		return fmt.Sprint(v)
	}
}
